{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import scipy as sp\n",
    "import numpy as np\n",
    "import multiprocessing as mp\n",
    "from sklearn.feature_selection import mutual_info_regression\n",
    "import seaborn as sns\n",
    "\n",
    "# Stockel Dataset\n",
    "\n",
    "df_stockel = pd.read_csv('MicroarrayData/StockelProcessed.csv')\n",
    "duplicates = df_stockel.loc[df_stockel.duplicated(subset='ORF',keep=False)].sort_values(by=['ORF'])\n",
    "mean_columns = list(duplicates.columns)\n",
    "mean_columns.remove('Contig')\n",
    "mean_columns.remove('ORF')\n",
    "df_expression = df_stockel.groupby('ORF')[mean_columns].mean().reset_index()\n",
    "\n",
    "# Genome Dataset\n",
    "\n",
    "GenCyanoDB = pd.read_excel('GenCyanoDB.xlsx',index_col=0,usecols=[0,1,2,3])\n",
    "\n",
    "# Merge the two DataFrames by their ORF column\n",
    "\n",
    "df = df_expression.merge(GenCyanoDB,on='ORF',how='inner')\n",
    "\n",
    "# Get the 2-fold change data\n",
    "\n",
    "filename = 'Pakrasi PatN Nif H analysis file.xlsx'\n",
    "filepath = '/Users/dzb5732/Box Sync/PhD-dzb5732@psu.edu/Cyanothece/Gene Expression Dataset/UWash-Raw/'\n",
    "file = filepath + filename\n",
    "df_2_fold = pd.read_excel(file,sheet_name='d4',usecols=[0,1,2,3,4,5])\n",
    "\n",
    "\n",
    "df_2_fold_nifH = df_2_fold.iloc[:,0:2]\n",
    "df_2_fold_nifH.columns = ['orf','fold_change']\n",
    "\n",
    "df_2_fold_patB = df_2_fold.iloc[:,[0,2]]\n",
    "df_2_fold_patB.columns = ['orf','fold_change']\n",
    "\n",
    "# Get Mutual Information\n",
    "\n",
    "class Interaction:\n",
    "    def __init__(self,Exp_data,gene='all',mi_thresh=0):\n",
    "        self.Exp_data = Exp_data\n",
    "        if self.Exp_data.isnull().values.any():\n",
    "            self.Exp_df = self.Exp_data.iloc[:,:-2].set_index('ORF').interpolate(method='linear',axis=1,limit_direction='both').T\n",
    "        else:\n",
    "            self.Exp_df = self.Exp_data.iloc[:,:-2].set_index('ORF').T\n",
    "        if gene=='all':\n",
    "            self.mi_dict = self._get_dict()\n",
    "        else:\n",
    "            self.gene_orf = gene\n",
    "            self.mi_list = self._miscorelist(self.gene_orf)\n",
    "            self.mi_thresh = mi_thresh\n",
    "            self.df = self._get_df(self.mi_list,self.mi_thresh)\n",
    "           \n",
    "    \n",
    "    def _get_dict(self):\n",
    "        all_genes = list(self.Exp_df.columns)\n",
    "        pool = mp.Pool(mp.cpu_count())\n",
    "        results = pool.map(self._miscorelist,all_genes)\n",
    "        fast_dict= dict(zip(all_genes,results))\n",
    "        return fast_dict\n",
    "\n",
    "    \n",
    "    def _miscorelist(self,gene):\n",
    "        all_other_genes_df = self.Exp_df.loc[:,self.Exp_df.columns!=gene]\n",
    "        all_other_genes = np.array(all_other_genes_df.columns)\n",
    "        this_gene_df = self.Exp_df[gene]\n",
    "        mi_score = mutual_info_regression(all_other_genes_df,this_gene_df,discrete_features=False,random_state=7)\n",
    "        miscore_genes = list(zip(all_other_genes,mi_score))\n",
    "        sorted_miscore = sorted(miscore_genes,key = lambda x:x[1],reverse=True)\n",
    "        return sorted_miscore\n",
    "    \n",
    "    def _get_df(self,mi_list,mi_thresh):\n",
    "        my_dict = {'orf':[],'function':[],'CommonName':[],'mi':[]}\n",
    "        for orf,mi in mi_list:\n",
    "            if mi<=mi_thresh:\n",
    "                break\n",
    "\n",
    "            my_dict['orf'].append(orf)\n",
    "            my_dict['function'].append(self.Exp_data.loc[self.Exp_data.ORF==orf].Function.values[0])\n",
    "            my_dict['CommonName'].append(self.Exp_data.loc[self.Exp_data.ORF==orf].CommonName.values[0])\n",
    "            my_dict['mi'].append(mi)\n",
    "\n",
    "        return pd.DataFrame(my_dict)\n",
    "\n",
    "## patB\n",
    "\n",
    "patB = Interaction(df,'cce_1898',-1)\n",
    "\n",
    "patB_info = df_2_fold_patB.merge(patB.df,how='outer',on='orf')\n",
    "\n",
    "## nifH\n",
    "\n",
    "nifH = Interaction(df,'cce_0559',-1)\n",
    "\n",
    "nifH_info = df_2_fold_nifH.merge(nifH.df,how='outer',on='orf')\n",
    "\n",
    "# Helper Functions\n",
    "\n",
    "def top_fold_changed_genes(df,direction='negative',top=25):\n",
    "    df.dropna(inplace=True)\n",
    "    return df.sort_values(by='fold_change',ascending=True).iloc[0:top] if direction=='negative' else df.sort_values(by='fold_change',ascending=False).iloc[0:top]\n",
    "\n",
    "def get_regulators(df,direction='negative'):\n",
    "    df.dropna(inplace=True)\n",
    "    return df.loc[abs(df.function.str.contains('regulator')) & (df.fold_change<-1)].sort_values(by='fold_change',ascending=True) if direction=='negative' else df.loc[abs(df.function.str.contains('regulator')) & (df.fold_change>1)].sort_values(by='fold_change',ascending=False) \n",
    "\n",
    "def get_sensors(df,direction='negative'):\n",
    "    df.dropna(inplace=True)\n",
    "    return df.loc[abs(df.function.str.contains('sensor')) & (df.fold_change<-1)].sort_values(by='fold_change',ascending=True) if direction=='negative' else df.loc[abs(df.function.str.contains('sensor')) & (df.fold_change>1)].sort_values(by='fold_change',ascending=False) \n",
    "\n",
    "def get_circadian(df,direction='negative'):\n",
    "    df.dropna(inplace=True)\n",
    "    return df.loc[abs(df.function.str.contains('circadian')) & (df.fold_change<-1)].sort_values(by='fold_change',ascending=True) if direction=='negative' else df.loc[abs(df.function.str.contains('circadian')) & (df.fold_change>1)].sort_values(by='fold_change',ascending=False) \n",
    "\n",
    "def get_sigma_factors(df,direction='negative'):\n",
    "    df.dropna(inplace=True)\n",
    "    return df.loc[abs(df.function.str.contains('sigma')) & (df.fold_change<-1)].sort_values(by='fold_change',ascending=True) if direction=='negative' else df.loc[abs(df.function.str.contains('sigma')) & (df.fold_change>1)].sort_values(by='fold_change',ascending=False) \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
